systemd-analyze compare-versions

keszybz opened the pull request.

bluca commented on src/test/test-string-util.c:
```diff
@@ -868,43 +879,187 @@ TEST(strverscmp_improved) {
+
+TEST(strverscmp_improved_rpm) {
+        /* Tests copied from rmp's rpmio test suite.
+         * The original form is retained for easy comparisons and updates. */
```
can you add also the link and the note about the license from the commit message? easier to see it as a comment

keszybz replied:
Done.

bluca commented on man/systemd-analyze.xml:
```diff
@@ -118,6 +118,13 @@
+    <cmdsynopsis>
+      <command>systemd-analyze</command>
+      <arg choice="opt" rep="repeat">OPTIONS</arg>
+      <arg choice="plain">vercmp</arg>
```
Nit: other verbs use the long format for the name rather than abbreviating, so what about `version-compare` for internal consistency?

keszybz replied:
So… I made it `vercmp` to match `rpmdev-vercmp`, but I'm fine with the longer version too if people prefer that.

bluca replied:
If we want this to be part of our public api, and part of the image story, then to me it seems better to go for internal consistency - but I do not have a strong opinion, just an observation

bluca commented on man/systemd-analyze.xml:
```diff
@@ -547,6 +554,31 @@ NAutoVTs=8
+      or <literal><replaceable>VERSION1</replaceable> &gt; <replaceable>VERSION2</replaceable></literal>
+      as appropriate.</para>
+
+      <para>The exit status is <constant>0</constant> if the versions are equal, <constant>11</constant> if
```
So not sure we need it, but `dpkg --compare-versions` has a nice interface because it allows the user to specify what the expected result should be, so it's really nice and flexible, eg:

```
$ dpkg --compare-versions 3.2.1.0 ge 0.1.2.3; echo $?
0
$ dpkg --compare-versions 3.2.1.0 lt 0.1.2.3; echo $?
1
```

Do we maybe want the same here?

bluca commented:
I know it's a very thin wrapper around the function which already has a unit test, but could you please add some minimal exercise of the binary in `test/units/testsuite-65.sh`? It should be really quick to do, and would cover the return code, which is part of the interface

keszybz commented:
Updated. I renamed the verb to `compare-versions`, and the interfaces of both `rpmdev-vercmp` and `dpkg --compare-versions` are now supported. This should make it easy for people to switch over to our implementation. The test is added as requested.

bluca commented:
> Updated. I renamed the verb to `compare-versions`, and the interfaces of both `rpmdev-vercmp` and `dpkg --compare-versions` are now supported. This should make it easy for people to switch over to our implementation. The test is added as requested.

Nice, thanks!

keszybz commented:
mkosi / ci (arch, rolling) failed because the internet is broken.

dtardon commented on man/systemd-analyze.xml:
```diff
@@ -1197,8 +1251,12 @@ $ systemd-analyze verify /tmp/source:alias.service
+    <para>With the verb <command>compare-versions</command>, in the two-argument form,
+    <constant>12</constant>, <constant>0</constant>, <constant>11</constant> is returned if the second
+    version string is respectively larger, equal, or smaller to the first. In the three-argument form,
+    <constant>0</constant> or <constant>0</constant> if the condition is respectively true or false.</para>
```
Here too, `<constant>1</constant>` for failure?

poettering commented:
I think we should also rework the function to not use locale dependent calls anymore, so that it has the same results everywhere.

i.e. isdigit() and isalpha() are all locale dependent. we should implement simpler replacements that just care about 7bit ASCII and nothing else.

poettering commented on src/analyze/analyze-capability.c:
```diff
@@ -48,5 +48,9 @@ int verb_capabilities(int argc, char *argv[], void *userdata) {
+        if (r < 0)
+                return r;
+
+        return 0;
```
ditto (and everywhere else).

i.e. if this is propagated as is now if > 0 we should probably use only EXIT_xyz names for returning here

poettering commented on src/analyze/analyze-compare-versions.c:
```diff
@@ -0,0 +1,45 @@
+                r = strverscmp_improved(ASSERT_PTR(argv[1]), ASSERT_PTR(argv[2]));
+                printf("%s %s %s\n",
+                       argv[1],
+                       r < 0 ? "<" : r > 0 ? ">" : "==",
```
Also, there's some code to share here with ConditionKernelVersion= i think, i.e. the OrderOperator enum

keszybz commented:
Updated. There's some more commits now, so a re-review is needed.

bluca commented:
New one?

```
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch -f
+ /usr/lib/rpm/rpmuncompress /builddir/build/SOURCES/f58b96d3e8d1cb0dd3666bc74fa673918b586612.patch
+ /usr/lib/rpm/rpmuncompress /builddir/build/SOURCES/0001-Revert-meson-create-new-libsystemd-core.so-private-s.patch
+ /usr/bin/patch -p1 -s --fuzz=0 --no-backup-if-mismatch -f
1 out of 2 hunks FAILED -- saving rejects to file meson.build.rej
```

keszybz commented:
/packit build

poettering commented:
What about sharing the parser with ConditionKernelVersion?

keszybz commented:
> What about sharing the parser with ConditionKernelVersion?

The important part is parsing of the version, which is shared, because both paths use `strverscmp_improved()`.
For the rest, we could try to share the part that extracts the operator, but I think that'd be more trouble than it's worth. The check is completely trivial here, because we get the operator as a separate item in argv[], at a known position, and the interpretation of the operator is also very simple. So I think this is fine as is.

poettering commented on src/fundamental/string-util-fundamental.c:
```diff
@@ -77,19 +77,18 @@ sd_char* endswith_no_case(const sd_char *s, const sd_char *postfix) {
 
-#ifdef SD_BOOT
-static sd_bool isdigit(sd_char a) {
+static sd_bool is_digit(sd_char a) {
```
maybe we should use a more explicit name. The underscore is not a good indicator that this is strictly doing ASCII stuff. hence, following the logic of ascii_toupper(), ascii_strlower(), ascii_strcasecmp(), we should call this helper ascii_isdigit() I think.

poettering replied:
and probably export it, in fact.
